bspwm.1
.TH "bspwm" 1 "June 23, 2013" "Bspwm"
.SH NAME
bspwm - Tiling window manager based on binary space partitioning
.PP
.SH SYNOPSIS
.PP
\f[B]bspwm\f[] [\f[B]\-h\f[]|\f[B]\-v\f[]|\f[B]\-s\f[] \f[I]PANEL_FIFO\f[]|\f[B]\-p\f[] \f[I]PANEL_PREFIX\f[]]
.PP
\f[B]bspc\f[] \f[I]MESSAGE\f[] [\f[I]ARGUMENTS\f[]] [\f[I]OPTIONS\[]]
.SH DESCRIPTION
.PP
\f[B]bspwm\f[] is a tiling window manager that represents windows as the leaves of a full binary tree.
.PP
It is controlled and configured via \f[B]bspc\f[].
.SH CONFIGURATION
.PP
\f[B]bspwm\f[] have only two sources of informations: the X events it receives and the messages it reads on a dedicated socket.
.PP
Its configuration file is \f[I]$XDG_CONFIG_HOME/bspwm/autostart\f[].
.PP
Keyboard and pointer bindings are defined with sxhkd.
.PP
Example configuration files can be found in the \f[B]examples\f[] directory.
.SH SPLITTING MODES
.PP
There is only two splitting modes: \f[I]automatic\f[] and \f[I]manual\f[].
.PP
The default mode is \f[I]automatic\f[].
The \f[I]manual\f[] mode is entered by sending a \f[B]presel\f[] message.
.PP
Example: insertion of a new node (number 4) into the given tree in \f[I]automatic\f[] mode:
.IP
.nf
\f[C]
\ \ \ \ \ \ \ \ \ \ \ \ \ \ b\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ c
\ \ \ \ \ \ \ \ \ \ \ \ \ /\ \\\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ /\ \\
\ \ \ \ \ \ \ \ \ \ \ \ 3\ \ \ a\ \ \ \ \ \ \ \ \ \ \ \ \ \ \-\->\ \ \ \ \ \ \ \ \ \ \ \ \ \ 4\ \ \ b
\ \ \ \ \ \ \ \ \ \ \ \ ^\ \ /\ \\\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ^\ \ /\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ 2\ \ \ 1\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 3\ \ \ a
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ /\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 2\ \ \ 1
\ +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\ \ \ \ \ \ \ \ \ +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
\ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |
\ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ 2\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ 3\ \ \ \ \ \ |
\ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |
\ |\ \ \ \ \ 3\ \ \ \ \ \ |\-\-\-\-\-\-\-\-\-\-\-\-|\ \ \ \-\->\ \ \ |\ \ \ \ \ 4\ \ \ \ \ \ |\-\-\-\-\-\-\-\-\-\-\-\-|
\ |\ \ \ \ \ ^\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ ^\ \ \ \ \ \ |\ \ \ \ \ |\ \ \ \ \ \ |
\ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ 1\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ 1\ \ |\ \ 2\ \ \ |
\ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ |\ \ \ \ \ \ |
\ +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\ \ \ \ \ \ \ \ \ +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
\f[]
.fi
.PP
Same departure, but the mode is \f[I]manual\f[], and a \f[B]presel\f[] \f[I]up\f[] message was sent beforehand:
.IP
.nf
\f[C]
\ \ \ \ \ \ \ \ \ \ \ \ \ b\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ b
\ \ \ \ \ \ \ \ \ \ \ \ /\ \\\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ /\ \\
\ \ \ \ \ \ \ \ \ \ \ 3\ \ \ a\ \ \ \ \ \ \ \ \ \ \ \ \ \ \-\->\ \ \ \ \ \ \ \ \ \ \ \ \ \ c\ \ \ a
\ \ \ \ \ \ \ \ \ \ \ ^\ \ /\ \\\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ /\ \\\ /\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ 2\ \ \ 1\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 4\ \ 3\ 2\ \ 1
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ^
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\ \ \ \ \ \ \ \ \ +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
|\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |
|\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ 2\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ 4\ \ \ \ \ \ |\ \ \ \ \ 2\ \ \ \ \ \ |
|\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ ^\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |
|\ \ \ \ \ 3\ \ \ \ \ \ |\-\-\-\-\-\-\-\-\-\-\-\-|\ \ \ \-\->\ \ \ |\-\-\-\-\-\-\-\-\-\-\-\-|\-\-\-\-\-\-\-\-\-\-\-\-|
|\ \ \ \ \ ^\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |
|\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ 1\ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ 3\ \ \ \ \ \ |\ \ \ \ \ 1\ \ \ \ \ \ |
|\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |\ \ \ \ \ \ \ \ \ \ \ \ |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\ \ \ \ \ \ \ \ \ +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
\f[]
.fi
.SH CONTAINERS
.PP
Each monitor contains at least one desktop.
.PP
Each desktop contains at most one tree.
.SH MESSAGES
.TP
.B \f[B]get\f[] \f[I]SETTING\f[]
Return the value of the given setting.
.RS
.RE
.TP
.B \f[B]set\f[] \f[I]SETTING\f[] \f[I]VALUE\f[]
Set the value of the given setting.
.RS
.RE
.TP
.B \f[B]list\f[] [\f[I]DESKTOP_NAME\f[]]
Output the internal representation of the window tree.
.RS
.RE
.TP
.B \f[B]list_desktops\f[] [\f[B]\-\-quiet\f[]]
Perform a dump of each desktop for the current monitor.
.RS
.RE
.TP
.B \f[B]list_monitors\f[] [\f[B]\-\-quiet\f[]]
Perform a dump of each monitor.
.RS
.RE
.TP
.B \f[B]list_history\f[]
Return the node focus history of each desktop.
.RS
.RE
.TP
.B \f[B]list_windows\f[]
Return the list of managed windows (i.e.
their identifiers).
.RS
.RE
.TP
.B \f[B]list_rules\f[]
Return the list of rules.
.RS
.RE
.TP
.B \f[B]presel\f[] \f[I]left\f[]|\f[I]right\f[]|\f[I]up\f[]|\f[I]down\f[] [\f[I]SPLIT_RATIO\f[]]
Switch to manual mode and select the splitting direction.
.RS
.RE
.TP
.B \f[B]cancel\f[]
Switch to automatic mode.
.RS
.RE
.TP
.B \f[B]ratio\f[] \f[I]VALUE\f[]
Set the splitting ratio of the focused window.
.RS
.RE
.TP
.B \f[B]pad\f[] \f[I]MONITOR_NAME\f[] [\f[I]TOP_PADDING\f[] [\f[I]RIGHT_PADDING\f[] [\f[I]BOTTOM_PADDING\f[] [\f[I]LEFT_PADDING\f[]]]]]
Set the padding of the given monitor.
.RS
.RE
.TP
.B \f[B]focus\f[] \f[I]left\f[]|\f[I]right\f[]|\f[I]up\f[]|\f[I]down\f[]
Focus the neighbor window situated in the given direction.
.RS
.RE
.TP
.B \f[B]shift\f[] \f[I]left\f[]|\f[I]right\f[]|\f[I]up\f[]|\f[I]down\f[]
Exchange the current window with the given neighbor.
.RS
.RE
.TP
.B \f[B]swap\f[] [\f[B]\-\-keep\-focus\f[]]
Swap the focused window with the last focused window.
.RS
.RE
.TP
.B \f[B]push\f[] \f[I]left\f[]|\f[I]right\f[]|\f[I]up\f[]|\f[I]down\f[]
Push the fence located in the given direction.
.RS
.RE
.TP
.B \f[B]pull\f[] \f[I]left\f[]|\f[I]right\f[]|\f[I]up\f[]|\f[I]down\f[]
Pull the fence located in the given direction.
.RS
.RE
.TP
.B \f[B]fence_ratio\f[] \f[I]left\f[]|\f[I]right\f[]|\f[I]up\f[]|\f[I]down\f[]
Set the splitting ratio of the fence located in the given direction.
.RS
.RE
.TP
.B \f[B]cycle\f[] \f[I]next\f[]|\f[I]prev\f[] [\f[B]\-\-skip\-floating\f[]|\f[B]\-\-skip\-tiled\f[]|\f[B]\-\-skip\-class\-equal\f[]|\f[B]\-\-skip\-class\-differ\f[]]
Focus the next or previous window matching the given constraints.
.RS
.RE
.TP
.B \f[B]nearest\f[] \f[I]older\f[]|\f[I]newer\f[] [\f[B]\-\-skip\-floating\f[]|\f[B]\-\-skip\-tiled\f[]|\f[B]\-\-skip\-class\-equal\f[]|\f[B]\-\-skip\-class\-differ\f[]]
Focus the nearest window matching the given constraints.
.RS
.RE
.TP
.B \f[B]biggest\f[]
Return the ID of the biggest tiled window.
.RS
.RE
.TP
.B \f[B]circulate\f[] \f[I]forward\f[]|\f[I]backward\f[]
Circulate the leaves in the given direction.
.RS
.RE
.TP
.B \f[B]grab_pointer\f[] \f[I]focus\f[]|\f[I]move\f[]|\f[I]resize_side\f[]|\f[I]resize_corner\f[]
Begin the specified pointer action.
.RS
.RE
.TP
.B \f[B]track_pointer\f[] \f[I]ROOT_X\f[] \f[I]ROOT_Y\f[]
Pass the pointer root coordinates for the current pointer action.
.RS
.RE
.TP
.B \f[B]ungrab_pointer\f[]
End the current pointer action.
.RS
.RE
.TP
.B \f[B]toggle_fullscreen\f[]
Toggle the fullscreen state of the current window.
.RS
.RE
.TP
.B \f[B]toggle_floating\f[]
Toggle the floating state of the current window.
.RS
.RE
.TP
.B \f[B]toggle_locked\f[]
Toggle the locked state of the current window (locked windows will not respond to the \f[B]close\f[] message).
.RS
.RE
.TP
.B \f[B]toggle_visibility\f[]
Toggle the visibility of all the managed windows.
.RS
.RE
.TP
.B \f[B]close\f[]
Close the focused window.
.RS
.RE
.TP
.B \f[B]kill\f[]
Kill the focused window.
.RS
.RE
.TP
.B \f[B]send_to\f[] \f[I]DESKTOP_NAME\f[] [\f[B]\-\-follow\f[]]
Send the focused window to the given desktop.
.RS
.RE
.TP
.B \f[B]drop_to\f[] \f[I]next\f[]|\f[I]prev\f[] [\f[B]\-\-follow\f[]]
Send the focused window to the next or previous desktop.
.RS
.RE
.TP
.B \f[B]send_to_monitor\f[] \f[I]MONITOR_NAME\f[] [\f[B]\-\-follow\f[]]
Send the focused window to the given monitor.
.RS
.RE
.TP
.B \f[B]drop_to_monitor\f[] \f[I]next\f[]|\f[I]prev\f[] [\f[B]\-\-follow\f[]]
Send the focused window to the next or previous monitor.
.RS
.RE
.TP
.B \f[B]use\f[] \f[I]DESKTOP_NAME\f[]
Select the given desktop.
.RS
.RE
.TP
.B \f[B]use_monitor\f[] \f[I]MONITOR_NAME\f[]
Select the given monitor.
.RS
.RE
.TP
.B \f[B]alternate\f[]
Alternate between the current and the last focused window.
.RS
.RE
.TP
.B \f[B]alternate_desktop\f[]
Alternate between the current and the last focused desktop.
.RS
.RE
.TP
.B \f[B]alternate_monitor\f[]
Alternate between the current and the last focused monitor.
.RS
.RE
.TP
.B \f[B]add\f[] \f[I]DESKTOP_NAME\f[] ...
Make new desktops with the given names.
.RS
.RE
.TP
.B \f[B]add_in\f[] \f[I]MONITOR_NAME\f[] \f[I]DESKTOP_NAME\f[] ...
Make new desktops with the given names in the given monitor.
.RS
.RE
.TP
.B \f[B]rename_monitor\f[] \f[I]CURRENT_NAME\f[] \f[I]NEW_NAME\f[]
Rename the monitor named \f[I]CURRENT_NAME\f[] to \f[I]NEW_NAME\f[].
.RS
.RE
.TP
.B \f[B]rename\f[] \f[I]CURRENT_NAME\f[] \f[I]NEW_NAME\f[]
Rename the desktop named \f[I]CURRENT_NAME\f[] to \f[I]NEW_NAME\f[].
.RS
.RE
.TP
.B \f[B]remove_desktop\f[] \f[I]DESKTOP_NAME\f[] ...
Remove the given desktops.
.RS
.RE
.TP
.B \f[B]send_desktop_to\f[] \f[I]MONITOR_NAME\f[] [\f[B]\-\-follow\f[]]
Send the current desktop to the given monitor.
.RS
.RE
.TP
.B \f[B]cycle_monitor\f[] \f[I]next\f[]|\f[I]prev\f[]
Select the next or previous monitor.
.RS
.RE
.TP
.B \f[B]cycle_desktop\f[] \f[I]next\f[]|\f[I]prev\f[] [\f[B]\-\-skip\-free\f[]|\f[B]\-\-skip\-occupied\f[]]
Select the next or previous desktop.
.RS
.RE
.TP
.B \f[B]layout\f[] \f[I]monocle\f[]|\f[I]tiled\f[] [\f[I]DESKTOP_NAME\f[] ...]
Set the layout of the given desktops (current if none given).
.RS
.RE
.TP
.B \f[B]cycle_layout\f[]
Cycle the layout of the current desktop.
.RS
.RE
.TP
.B \f[B]rotate\f[] \f[I]clockwise\f[]|\f[I]counter_clockwise\f[]|\f[I]full_cycle\f[]
Rotate the window tree.
.RS
.RE
.TP
.B \f[B]flip\f[] \f[I]horizontal\f[]|\f[I]vertical\f[]
Flip the window tree.
.RS
.RE
.TP
.B \f[B]balance\f[]
Adjust the split ratios so that all windows occupy the same area.
.RS
.RE
.TP
.B \f[B]rule\f[] \f[I]PATTERN\f[] [\f[I]DESKTOP_NAME\f[]] [\f[I]floating\f[]] [\f[I]follow\f[]]
Create a new rule (\f[I]PATTERN\f[] must match the class or instance name).
.RS
.RE
.TP
.B \f[B]remove_rule\f[] \f[I]UID\f[] ...
Remove the rules with the given \f[I]UID\f[]s.
.RS
.RE
.TP
.B \f[B]put_status\f[]
Output the current state to the panel fifo.
.RS
.RE
.TP
.B \f[B]adopt_orphans\f[]
Manage all the unmanaged windows remaining from a previous session.
.RS
.RE
.TP
.B \f[B]restore_layout\f[] \f[I]FILE_PATH\f[]
Restore the layout of each desktop from the content of \f[I]FILE_PATH\f[].
.RS
.RE
.TP
.B \f[B]restore_history\f[] \f[I]FILE_PATH\f[]
Restore the history of each desktop from the content of \f[I]FILE_PATH\f[].
.RS
.RE
.TP
.B \f[B]quit\f[] [\f[I]EXIT_STATUS\f[]]
Quit.
.RS
.RE
.SH SETTINGS
.PP
Colors are either X color names or \f[I]#RRGGBB\f[], booleans are \f[I]true\f[] or \f[I]false\f[].
.PP
All the boolean settings are \f[I]false\f[] by default.
.TP
.B \f[I]focused_border_color\f[]
Color of the border of a focused window of a focused monitor.
.RS
.RE
.TP
.B \f[I]active_border_color\f[]
Color of the border of a focused window of an unfocused monitor.
.RS
.RE
.TP
.B \f[I]normal_border_color\f[]
Color of the border of an unfocused window.
.RS
.RE
.TP
.B \f[I]presel_border_color\f[]
Color of the \f[B]presel\f[] message feedback.
.RS
.RE
.TP
.B \f[I]focused_locked_border_color\f[]
Color of the border of a focused locked window of a focused monitor.
.RS
.RE
.TP
.B \f[I]active_locked_border_color\f[]
Color of the border of a focused locked window of an unfocused monitor.
.RS
.RE
.TP
.B \f[I]normal_locked_border_color\f[]
Color of the border of an unfocused locked window.
.RS
.RE
.TP
.B \f[I]urgent_border_color\f[]
Color of the border of an urgent window.
.RS
.RE
.TP
.B \f[I]border_width\f[]
Window border width.
.RS
.RE
.TP
.B \f[I]window_gap\f[]
Value of the gap that separates windows.
.RS
.RE
.TP
.B \f[I]split_ratio\f[]
Default split ratio.
.RS
.RE
.TP
.B \f[I]top_padding\f[], \f[I]right_padding\f[], \f[I]bottom_padding\f[], \f[I]left_padding\f[]
Padding space added at the sides of the current monitor.
.RS
.RE
.TP
.B \f[I]wm_name\f[]
The value that shall be used for the \f[I]_NET_WM_NAME\f[] property of the root window.
.RS
.RE
.TP
.B \f[I]borderless_monocle\f[]
Remove borders for tiled windows in monocle mode.
.RS
.RE
.TP
.B \f[I]gapless_monocle\f[]
Remove gaps for tiled windows in monocle mode.
.RS
.RE
.TP
.B \f[I]focus_follows_pointer\f[]
Focus the window under the pointer.
.RS
.RE
.TP
.B \f[I]pointer_follows_monitor\f[]
When focusing a monitor, put the pointer at its center.
.RS
.RE
.TP
.B \f[I]adaptative_raise\f[]
Prevent floating windows from being raised when they might cover other floating windows.
.RS
.RE
.TP
.B \f[I]apply_shadow_property\f[]
Enable shadows for floating windows via the \f[I]_COMPTON_SHADOW\f[] property.
.RS
.RE
.TP
.B \f[I]auto_alternate\f[]
Interpret two consecutive identical \f[B]use\f[] messages as an \f[B]alternate\f[] message.
.RS
.RE
.TP
.B \f[I]focus_by_distance\f[]
Use window or leaf distance for focus movement.
.RS
.RE
.TP
.B \f[I]history_aware_focus\f[]
Give priority to the focus history when focusing nodes.
.RS
.RE
.SH ENVIRONMENT VARIABLES
.TP
.B \f[I]BSPWM_SOCKET\f[]

bspwm.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <xcb/xcb.h>
#include <xcb/xcb_event.h>
#include <xcb/xcb_ewmh.h>
#include <xcb/randr.h>
#include "types.h"
#include "settings.h"
#include "messages.h"
#include "rules.h"
#include "events.h"
#include "common.h"
#include "helpers.h"
#include "window.h"
#include "bspwm.h"
#include "tree.h"
#include "ewmh.h"

void quit(void)
{
    running = false;
}

void cleanup(void)
{
    while (mon_head != NULL)
        remove_monitor(mon_head);
    while (rule_head != NULL)
        remove_rule(rule_head);
    free(frozen_pointer);
}

void register_events(void)
{
    uint32_t values[] = {ROOT_EVENT_MASK};
    xcb_generic_error_t *e = xcb_request_check(dpy, xcb_change_window_attributes_checked(dpy, root, XCB_CW_EVENT_MASK, values));
    if (e != NULL) {
        xcb_disconnect(dpy);
        err("Another window manager is already running.\n");
    }
}

bool import_monitors(void)
{
    PUTS("import monitors");
    xcb_randr_get_screen_resources_current_reply_t *sres = xcb_randr_get_screen_resources_current_reply(dpy, xcb_randr_get_screen_resources_current(dpy, root), NULL);
    if (sres == NULL)
        return false;

    int len = xcb_randr_get_screen_resources_current_outputs_length(sres);
    xcb_randr_output_t *outputs = xcb_randr_get_screen_resources_current_outputs(sres);

    xcb_randr_get_output_info_cookie_t cookies[len];
    for (int i = 0; i < len; i++)
        cookies[i] = xcb_randr_get_output_info(dpy, outputs[i], XCB_CURRENT_TIME);

    for (monitor_t *m = mon_head; m != NULL; m = m->next)
        m->wired = false;

    monitor_t *mm = NULL;
    unsigned int num = 0;

    for (int i = 0; i < len; i++) {
        xcb_randr_get_output_info_reply_t *info = xcb_randr_get_output_info_reply(dpy, cookies[i], NULL);
        if (info != NULL && info->crtc != XCB_NONE) {

            xcb_randr_get_crtc_info_reply_t *cir = xcb_randr_get_crtc_info_reply(dpy, xcb_randr_get_crtc_info(dpy, info->crtc, XCB_CURRENT_TIME), NULL);
            if (cir != NULL) {
                xcb_rectangle_t rect = (xcb_rectangle_t) {cir->x, cir->y, cir->width, cir->height};
                mm = get_monitor_by_id(outputs[i]);
                if (mm != NULL) {
                    mm->rectangle = rect;
                    for (desktop_t *d = mm->desk_head; d != NULL; d = d->next)
                        for (node_t *n = first_extrema(d->root); n != NULL; n = next_leaf(n, d->root))
                            fit_monitor(mm, n->client);
                    arrange(mm, mm->desk);
                    mm->wired = true;
                    PRINTF("update monitor %s (0x%X)\n", mm->name, mm->id);
                } else {
                    mm = add_monitor(&rect);
                    char *name = (char *)xcb_randr_get_output_info_name(info);
                    size_t name_len = MIN(sizeof(mm->name), (size_t)xcb_randr_get_output_info_name_length(info));
                    strncpy(mm->name, name, name_len);
                    mm->name[name_len] = '\0';
                    mm->id = outputs[i];
                    add_desktop(mm, make_desktop(NULL));
                    PRINTF("add monitor %s (0x%X)\n", mm->name, mm->id);
                }
                num++;
            }
            free(cir);
        }
        free(info);
    }

    monitor_t *m = mon_head;
    while (m != NULL) {
        monitor_t *next = m->next;
        if (!m->wired) {
            PRINTF("remove monitor %s (0x%X)\n", m->name, m->id);
            merge_monitors(m, mm);
            remove_monitor(m);
        }
        m = next;
    }

    free(sres);
    update_motion_recorder();
    return (num_monitors > 0);
}

void init(void)
{
    num_monitors = num_desktops = num_clients = 0;
    monitor_uid = desktop_uid = client_uid = rule_uid = 0;
    mon = last_mon = mon_head = mon_tail = NULL;
    rule_head = rule_tail = NULL;
    randr_base = 0;
    visible = true;
    exit_status = 0;
}

void setup(void)
{
    init();
    ewmh_init();
    screen = xcb_setup_roots_iterator(xcb_get_setup(dpy)).data;
    if (screen == NULL)
        err("Can't acquire the default screen.\n");
    root = screen->root;
    register_events();

    screen_width = screen->width_in_pixels;
    screen_height = screen->height_in_pixels;
    root_depth = screen->root_depth;

    uint32_t mask = XCB_CW_EVENT_MASK;
    uint32_t values[] = {XCB_EVENT_MASK_POINTER_MOTION};
    motion_recorder = xcb_generate_id(dpy);
    xcb_create_window(dpy, XCB_COPY_FROM_PARENT, motion_recorder, root, 0, 0, screen_width, screen_height, 0, XCB_WINDOW_CLASS_INPUT_ONLY, XCB_COPY_FROM_PARENT, mask, values);

    xcb_atom_t net_atoms[] = {ewmh->_NET_SUPPORTED,
                              ewmh->_NET_DESKTOP_NAMES,
                              ewmh->_NET_NUMBER_OF_DESKTOPS,
                              ewmh->_NET_CURRENT_DESKTOP,
                              ewmh->_NET_CLIENT_LIST,
                              ewmh->_NET_ACTIVE_WINDOW,
                              ewmh->_NET_WM_DESKTOP,
                              ewmh->_NET_WM_STATE,
                              ewmh->_NET_WM_STATE_FULLSCREEN,
                              ewmh->_NET_WM_STATE_DEMANDS_ATTENTION,
                              ewmh->_NET_WM_WINDOW_TYPE,
                              ewmh->_NET_WM_WINDOW_TYPE_DOCK,
                              ewmh->_NET_WM_WINDOW_TYPE_NOTIFICATION,
                              ewmh->_NET_WM_WINDOW_TYPE_DIALOG,
                              ewmh->_NET_WM_WINDOW_TYPE_UTILITY,
                              ewmh->_NET_WM_WINDOW_TYPE_TOOLBAR};

    xcb_ewmh_set_supported(ewmh, default_screen, LENGTH(net_atoms), net_atoms);

    xcb_intern_atom_reply_t *iar = xcb_intern_atom_reply(dpy, xcb_intern_atom(dpy, 0, strlen("_COMPTON_SHADOW"), "_COMPTON_SHADOW"), NULL);

    if (iar != NULL) {
        compton_shadow = iar->atom;
        free(iar);
    }

    const xcb_query_extension_reply_t *qep = xcb_get_extension_data(dpy, &xcb_randr_id);
    if (qep->present && import_monitors()) {
        randr = true;
        randr_base = qep->first_event;
        xcb_randr_select_input(dpy, root, XCB_RANDR_NOTIFY_MASK_SCREEN_CHANGE);
    } else {
        randr = false;
        warn("Couldn't retrieve monitors via RandR.\n");
        xcb_rectangle_t rect = (xcb_rectangle_t) {0, 0, screen_width, screen_height};
        monitor_t *m = add_monitor(&rect);
        add_desktop(m, make_desktop(NULL));
    }

    ewmh_update_number_of_desktops();
    ewmh_update_desktop_names();
    ewmh_update_current_desktop();
    frozen_pointer = make_pointer_state();
}

int main(int argc, char *argv[])
{
    fd_set descriptors;
    char socket_path[MAXLEN];
    char *fifo_path = NULL;
    status_prefix = NULL;
    int sock_fd, ret_fd, dpy_fd, sel, n;
    struct sockaddr_un sock_address;
    size_t rsplen = 0;
    char msg[BUFSIZ] = {0};
    char rsp[BUFSIZ] = {0};
    xcb_generic_event_t *event;
    char opt;

    while ((opt = getopt(argc, argv, "hvs:p:")) != -1) {
        switch (opt) {
            case 'h':
                printf("bspwm [-h|-v|-s PANEL_FIFO|-p PANEL_PREFIX]\n");
                exit(EXIT_SUCCESS);
                break;
            case 'v':
                printf("%s\n", VERSION);
                exit(EXIT_SUCCESS);
                break;
            case 's':
                fifo_path = optarg;
                break;
            case 'p':
                status_prefix = optarg;
                break;
        }
    }

    running = true;
    dpy = xcb_connect(NULL, &default_screen);

    if (xcb_connection_has_error(dpy))
        err("Can't open the default display.\n");

    setup();

    dpy_fd = xcb_get_file_descriptor(dpy);

    char *sp = getenv(SOCKET_ENV_VAR);
    strncpy(socket_path, (sp == NULL ? DEFAULT_SOCKET_PATH : sp), sizeof(socket_path));

    sock_address.sun_family = AF_UNIX;
    strncpy(sock_address.sun_path, socket_path, sizeof(sock_address.sun_path));
    unlink(socket_path);

    sock_fd = socket(AF_UNIX, SOCK_STREAM, 0);

    if (sock_fd == -1)
        err("Couldn't create the socket.\n");

    if (bind(sock_fd, (struct sockaddr *) &sock_address, sizeof(sock_address)) == -1)
        err("Couldn't bind a name to the socket.\n");

    if (listen(sock_fd, SOMAXCONN) == -1)
        err("Couldn't listen to the socket.\n");

    sel = MAX(sock_fd, dpy_fd) + 1;

    if (fifo_path != NULL) {
        int fifo_fd = open(fifo_path, O_RDWR | O_NONBLOCK);
        if (fifo_fd != -1)
            status_fifo = fdopen(fifo_fd, "w");
        else
            warn("Couldn't open status fifo.\n");
    }

    load_settings();
    run_autostart();
    ewmh_update_wm_name();

    while (running) {

        xcb_flush(dpy);

        FD_ZERO(&descriptors);
        FD_SET(sock_fd, &descriptors);
        FD_SET(dpy_fd, &descriptors);

        if (select(sel, &descriptors, NULL, NULL, NULL) > 0) {

            if (FD_ISSET(sock_fd, &descriptors)) {
                ret_fd = accept(sock_fd, NULL, 0);
                if (ret_fd > 0 && (n = recv(ret_fd, msg, sizeof(msg), 0)) > 0) {
                    msg[n] = '\0';
                    process_message(msg, rsp);
                    rsplen = strlen(rsp);
                    if (rsp[rsplen - 1] == '\n')
                        rsp[--rsplen] = '\0';
                    send(ret_fd, rsp, rsplen, 0);
                    close(ret_fd);
                    rsp[0] = '\0';
                }
            }

            if (FD_ISSET(dpy_fd, &descriptors)) {
                while ((event = xcb_poll_for_event(dpy)) != NULL) {
                    handle_event(event);
                    free(event);
                }
            }

        }

        if (xcb_connection_has_error(dpy))
            err("The server has closed the connection.\n");
    }

    cleanup();
    close(sock_fd);
    if (status_fifo != NULL)
        fclose(status_fifo);
    xcb_ewmh_connection_wipe(ewmh);
    xcb_destroy_window(dpy, motion_recorder);
    free(ewmh);
    xcb_flush(dpy);
    xcb_disconnect(dpy);
    return exit_status;
}
